pkg/healthcheck/healthcheck.go

```diff
@@ -2191,6 +2209,90 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
 	return nil
 }
 
+// Check if there's a pod with the opaque-ports annotation defined but a
+// service selecting the aforementioned pod doesn't define it
+func (hc *HealthChecker) checkMisconfiguredOpaquePortAnnotations(ctx context.Context, namespace string) error {
+	services, err := hc.GetServices(ctx)
+	if err != nil {
+		return err
+	}
+
+	pods, err := hc.kubeAPI.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	var errStrings []string
+
+	for _, service := range services {
+		if mismatch := misconfiguredOpaquePortAnnotationsInService(service, pods); mismatch {
+			errStrings = append(
+				errStrings,
+				fmt.Sprintf("%s annotation is not properly configured in service %s",
+					k8s.ProxyOpaquePortsAnnotation,
+					service.Name),
+			)
+		}
+	}
+
+	if len(errStrings) >= 1 {
+		return fmt.Errorf(strings.Join(errStrings, "\n"))
+	}
+
+	return nil
+}
+
+func misconfiguredOpaquePortAnnotationsInService(service corev1.Service, pods *corev1.PodList) bool {
+	for _, pod := range pods.Items {
+		// The endpoints for headless services are handled differently so don't check this type of services
+		if service.Spec.ClusterIP != "None" && serviceMatchesPod(service, pod) {
```

Comment:
If the service's cluster IP is `None` then we'll never check any of it's pods, so it'd be good to move this check outside of the `pods` loop.

Additionally if we move forward with the selector change, we'll already know at this point which pods match the service, so the `serviceMatchesPod` check shouldn't be needed.